src/Stripe.net/Infrastructure/JsonConverters/StripeEntityConverter.cs

```diff
@@ -0,0 +1,83 @@
+﻿namespace Stripe.Infrastructure
+{
+    using System;
+    using System.IO;
+    using System.Runtime.Serialization;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// This converter is used to deserialize objects inheriting from StripeEntity.
+    /// It sets the RawJObject property so that undocumented and unsupported fields can be accessed or logged.
+    /// </summary>
+    public class StripeEntityConverter : JsonConverter
+    {
+        /// <summary>
+        /// Gets a value indicating whether this <see cref="JsonConverter"/> can write JSON.
+        /// </summary>
+        /// <value>
+        ///     <c>true</c> if this <see cref="JsonConverter"/> can write JSON; otherwise, <c>false</c>.
+        /// </value>
+        public override bool CanWrite => false;
+
+        /// <summary>
+        /// Determines whether this instance can convert the specified object type.
+        /// </summary>
+        /// <param name="objectType">Type of the object.</param>
+        /// <returns>
+        ///     <c>true</c> if this instance can convert the specified object type; otherwise, <c>false</c>.
+        /// </returns>
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(StripeEntity).IsAssignableFrom(objectType);
+        }
+
+        /// <summary>
+        /// Writes the JSON representation of the object.
+        /// </summary>
+        /// <param name="writer">The <see cref="JsonWriter"/> to write to.</param>
+        /// <param name="value">The value.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            throw new NotSupportedException($"{nameof(StripeEntityConverter)} should only be used while deserializing.");
+        }
+
+        /// <summary>
+        /// Reads the JSON representation of the object.
+        /// </summary>
+        /// <param name="reader">The <see cref="JsonReader"/> to read from.</param>
+        /// <param name="objectType">Type of the object.</param>
+        /// <param name="existingValue">The existing value of object being read.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        /// <returns>The object value.</returns>
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
+
+            JToken jToken;
+            if (reader is JTokenReader jTokenReader)
```

Comment:
Hm, I don't think we ever create a `JTokenReader` and the default implementation also doesn't use it: https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/JsonConvert.cs#L829

So we never hit the optimized codepath.

Reply:
I guess I should have added some comments.

When calling `jToken.CreateReader()`, it actually returns an `JTokenReader` internally. `JTokenReader` is more efficient than `JsonReader` because instead of parsing the raw text, it just visits the existing json tree.

What happens is that the top level entity is parsed normally when calling `JToken.Load(reader)`. However, nested entities will detect that the reader is a `JTokenReader` and will create a new sub reader using the existing jToken.
As a result, the json is parsed once only. In addition, all of the `RawObject` properties of all the nested entities reference a node from the single top level tree so the json is loaded in memory once only.

I hope that makes sense.

If you want proof that this works, I would suggest that you merge this in (or check out my branch), run the test below and put a breakpoint in the `StripeEntityConverter`.

https://github.com/stripe/stripe-dotnet/blob/7b62c461d7c0cf2c9e06dce5e564b374a9d232e0/src/StripeTests/Entities/Accounts/AccountTest.cs#L14-L18

Also, you can see that the other converters actually do already create `JTokenReader` whenever they create a subreader (with the same pattern as here).
https://github.com/stripe/stripe-dotnet/blob/7b62c461d7c0cf2c9e06dce5e564b374a9d232e0/src/Stripe.net/Infrastructure/JsonConverters/AnyOfConverter.cs#L70
https://github.com/stripe/stripe-dotnet/blob/7b62c461d7c0cf2c9e06dce5e564b374a9d232e0/src/Stripe.net/Infrastructure/JsonConverters/StripeObjectConverter.cs#L63
https://github.com/stripe/stripe-dotnet/blob/7b62c461d7c0cf2c9e06dce5e564b374a9d232e0/src/Stripe.net/Infrastructure/JsonConverters/ExpandableFieldConverter.cs#L76
https://github.com/stripe/stripe-dotnet/blob/7b62c461d7c0cf2c9e06dce5e564b374a9d232e0/src/Stripe.net/Infrastructure/JsonConverters/EventConverter.cs#L72

Reply:
Ah, you are right. I'm sorry I missed that.

Do you think there is any way to get rid of the Reader allocation that we now have for every entity?